native-dash-app-example/src/app.py

```diff
@@ -0,0 +1,217 @@
+API_URL = os.environ.get("JUPYTERHUB_API_URL", "http://127.0.0.1:8081")
+APP_VERSION = os.environ.get("APP_VERSION", "native-dash-app-example")
+EDGE_API_SERVICE_URL = os.environ.get("EDGE_API_SERVICE_URL", None)
+EDGE_API_ORG = os.environ.get("EDGE_API_ORG", None)
+
+DASH_URL_BASE_PATHNAME = "dashboard/"
+
+AUTH = HubOAuth(api_token=API_TOKEN, cache_max_age=60, api_url=API_URL)
+
+LOG.debug(f"JUPYTERHUB_SERVER_NAME {SERVER_NAME}")
+LOG.debug(f"JUPYTERHUB_SERVICE_PREFIX {PREFIX}")
+LOG.debug(f"JUPYTERHUB_ACTIVITY_URL {ACTIVITY_URL}")
+
+
+def track_activity(f):
+    """Decorator for reporting server activities with the Hub"""
+
+    @wraps(f)
+    def decorated(*args, **kwargs):
+        if FLASK_DEBUG == 1:
+            return f(*args, **kwargs)
+        last_activity = isoformat(datetime.datetime.now())
+        if ACTIVITY_URL:
+            try:
+                requests.post(
+                    ACTIVITY_URL,
+                    headers={
+                        "Authorization": f"token {API_TOKEN}",
+                        "Content-Type": "application/json",
+                    },
+                    json={"servers": {SERVER_NAME: {"last_activity": last_activity}}},
+                )
+            except Exception:
+                pass
+        return f(*args, **kwargs)
+
+    return decorated
+
+
+def view_authenticated(view_func, **view_args):
+    """Wraps the view function with authentication."""
+    token = session.get("token")
+    if token:
+        hub_user = AUTH.user_for_token(token)
+    else:
+        hub_user = None
+
+    if hub_user or session.get("redirecting") is True:
+        session["redirecting"] = False
+        return view_func(**view_args)
+    else:
+        # redirect to login url on failed auth
+        # Note that we are redirecting so we don't endlessly loop authn
+        session["redirecting"] = True
+        state = AUTH.generate_state(next_url=request.path)
+        LOG.info(f"Redirecting to login url {AUTH.login_url}")
+        response = make_response(redirect(AUTH.login_url + "&state=%s" % state))
+        response.set_cookie(AUTH.state_cookie_name, state)
+        return response
+
+
+def create_app():
+    """Creates the Flask app with routes for serving the React application
+    and API routes for running jobs
+    """
+    flask = Flask(
+        __name__,
+        template_folder="templates",
+        static_url_path=PREFIX + "static",
+    )
+    flask.config["SESSION_TYPE"] = "filesystem"
+    flask.config["SECRET_KEY"] = "super secret key"
+    sess = Session()
+    sess.init_app(flask)
+    flask.jinja_env.filters["url_decode"] = lambda url: unquote(url)
+
+    app = dash.Dash(
+        __name__,
+        server=flask,
+        url_base_pathname=PREFIX + DASH_URL_BASE_PATHNAME,
+    )
+
+    df = pd.read_csv(
+        "https://raw.githubusercontent.com/plotly/datasets/master/gapminder_unfiltered.csv"
+    )
+
+    app.layout = html.Div(
```

Comment:
`app.layout` can also be assigned a function that returns `html.Div` layouts, with auth performed before render.

Reply:
I will try it, but the question that leaps to mind is if it will handle the redirect the same as the view wrapper.

Reply:
It has all kinds of problems because the app.layout runs outside of a Flask requests context.

Reply:
I do notice you have a series of api endpoints to manage session and authentication in your oauth2 example, though I'm not sure how much work we want to put into this temporary dash example until we shift it to the oauth2 docker variation.